runtimebp/internal/maxprocs/maxprocs.go

```diff
@@ -0,0 +1,111 @@
+	setByGOMAXPROCS = "gomaxprocs"
+	setByRequest    = "cpu_request"
+)
+
+var (
+	mEnvGOMAXPROCS      = promauto.NewGaugeVec(prometheus.GaugeOpts{Name: "baseplate_go_env_gomaxprocs"}, []string{"present"})
+	mEnvCPURequest      = promauto.NewGaugeVec(prometheus.GaugeOpts{Name: "baseplate_go_env_baseplate_cpu_request"}, []string{"present"})
+	mEnvCPURequestScale = promauto.NewGaugeVec(prometheus.GaugeOpts{Name: "baseplate_go_env_baseplate_cpu_request_scale"}, []string{"present"})
+
+	initialGOMAXPROCS = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "baseplate_go_initial_gomaxprocs",
+	}, []string{"set_by"})
+
+	_ = promauto.NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "baseplate_go_current_gomaxprocs",
+	}, currentGOMAXPROCS)
+)
+
+// Set configures the runtime's GOMAXPROCS using the following heuristic:
+//   1. If $GOMAXPROCS is set, Set relinquishes control to the Go runtime.
+//      This should cause the runtime to respect this value directly.
+//   2. If $BASEPLATE_CPU_REQUEST is unset/invalid, Set relinquishes control to the Go runtime.
+//      This should cause the runtime to use the detected number of CPUs on this machine.
+//   3. Otherwise, $BASEPLATE_CPU_REQUEST is multiplied by $BASEPLATE_CPU_REQUEST_SCALE
+//      (or defaultCPURequestScale) to compute the new GOMAXPROCS, minimum 2.
+//
+// Set also exports several metrics to facilitate further tuning/analysis.
+func Set() {
+	envGOMAXPROCS := &floatEnv{key: "GOMAXPROCS", gauge: mEnvGOMAXPROCS}
+	envCPURequest := &floatEnv{key: "BASEPLATE_CPU_REQUEST", gauge: mEnvCPURequest}
+	envCPURequestScale := &floatEnv{key: "BASEPLATE_CPU_REQUEST_SCALE", gauge: mEnvCPURequestScale}
+
+	for _, env := range []*floatEnv{envGOMAXPROCS, envCPURequest, envCPURequestScale} {
+		env.raw, env.present = os.LookupEnv(env.key)
+		env.val, _ = strconv.ParseFloat(env.raw, 64)
+		env.gauge.WithLabelValues(strconv.FormatBool(env.present)).Set(env.val)
+	}
+
+	setBy := setByGOMAXPROCS
+	defer func() { initialGOMAXPROCS.WithLabelValues(setBy).Set(currentGOMAXPROCS()) }()
+
+	if envGOMAXPROCS.present {
+		return // let Go runtime handle it
+	}
+
+	if !envCPURequest.present {
+		return // let Go runtime handle it
+	}
+
+	if envCPURequest.val <= 0 {
+		// This should always be valid positive float in infrared-deployed applications.
+		fmt.Fprintf(os.Stderr, "maxprocs: $BASEPLATE_CPU_REQUEST=%q, want positive float. Falling back to Go's default", envCPURequest.raw)
+		return
+	}
+
+	scale := defaultCPURequestScale
+	if envCPURequestScale.val > 0 {
+		scale = envCPURequestScale.val
+	} else if envCPURequestScale.present {
+		fmt.Fprintf(os.Stderr, "maxprocs: $BASEPLATE_CPU_REQUEST_SCALE=%q, want positive float. Falling back to default of %g", envCPURequestScale.raw, scale)
+	}
+
+	setBy = setByRequest
+	runtime.GOMAXPROCS(int(
+		math.Max(
+			2, // to ensure some minimal parallelism
+			math.Ceil(envCPURequest.val*scale),
+		),
+	))
+}
+
+func currentGOMAXPROCS() float64 {
+	return float64(runtime.GOMAXPROCS(0))
```

Comment:
this feels like an abuse of the sideeffect of `runtime.GOMAXPROCS` to me. there must be a better way of getting this info?

Reply:
This is the only way to access the current GOMAXPROCS value as far as I know.

Reply:
There isn't.  This is exactly how you are intended to get GOMAXPROCS.